src/WallpaperEngine/Core/CWallpaper.h

```diff
@@ -0,0 +1,33 @@
+#pragma once
+
+#include <assert.h>
+
+#include "CProject.h"
+
+namespace WallpaperEngine::Core
+{
+    class CProject;
+
+    class CWallpaper
```

Comment:
To allow the usage of IrrlichtContext for the rendering the CWallpaper class must inherit 
> irr::scene::ISceneNode

This way you don't need to manually do the begin scene on the CVideo class, only generate the frame (on the render function of irr::scene::ISceneNode) and perform a draw2DImage (which should be bound to the viewport being rendered).

Reply:
I'm uncertain how to override the render function in derived classes of CWallpaper. Would I have the declare the virtual void render in CWallpaper, again?

Reply:
You don't need to re-declare the function in CWallpaper, just do the normal override in the derived classes where you want to give the render function custom functionality. (btw I've just realized I've commented this on the Core part of CWallpaper, it should be on the Render part of CWallpaper).